logstash-core/src/test/java/org/logstash/config/ir/CompiledPipelineTest.java

```diff
@@ -560,4 +564,130 @@ public Filter buildFilter(final String name, final String id,
             return null;
         }
     }
+
+    @Test
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    public void compilerBenchmark() throws Exception {
+        final PipelineIR baselinePipelineIR = createPipelineIR(200);
+        final PipelineIR testPipelineIR = createPipelineIR(400);
+        final JrubyEventExtLibrary.RubyEvent testEvent =
+                JrubyEventExtLibrary.RubyEvent.newRubyEvent(RubyUtil.RUBY, new Event());
+
+        final FixedPluginFactory pluginFactory = new FixedPluginFactory(
+                () -> null,
+                () -> IDENTITY_FILTER,
+                mockOutputSupplier()
+        );
+        final CompiledPipeline baselineCompiledPipeline = new CompiledPipeline(baselinePipelineIR, pluginFactory);
+
+        final CompiledPipeline testCompiledPipeline = new CompiledPipeline(testPipelineIR, pluginFactory);
+
+        final long compilationBaseline = time(ChronoUnit.SECONDS, () -> {
+            final CompiledPipeline.CompiledExecution compiledExecution = baselineCompiledPipeline.buildExecution();
+            compiledExecution.compute(RubyUtil.RUBY.newArray(testEvent), false, false);
+        });
+        System.out.println(String.format(">>>DNADBG compilation baseline took: %d seconds", compilationBaseline));
+
+        final long compilationTest = time(ChronoUnit.SECONDS, () -> {
+            final CompiledPipeline.CompiledExecution compiledExecution = testCompiledPipeline.buildExecution();
+            compiledExecution.compute(RubyUtil.RUBY.newArray(testEvent), false, false);
+        });
+        System.out.println(String.format(">>>DNADBG compilation test took: %d seconds", compilationTest));
+
+        // sanity checks
+        final Collection<JrubyEventExtLibrary.RubyEvent> outputEvents = EVENT_SINKS.get(runId);
+        MatcherAssert.assertThat(outputEvents.size(), CoreMatchers.is(2));
+        MatcherAssert.assertThat(outputEvents.contains(testEvent), CoreMatchers.is(true));
+
+        // regression check
+        final String testMessage = "regression in pipeline compilation, twice the filters require more than twice " +
```

Comment:
I think the tests here should not reflect on time taken to compile, but on the cache reuse.
The test should check that compiling two filters in the same position of the filter stage should be cached even if the ID is different.

Reply:
and just to be clear, we need these tests included in this PR, but I think they should belong in a performance regression test suite.
however this PR addresses an issue where classes could no longer be reused from the cache and that's what the test should hit on.

Reply:
Sounds reasonable, I ask why the "should check that compiling two filters in the same position of the filter stage", the caching happens on the generated code, so the position inside the pipeline  shouldn't have effect?

Reply:
the generated code still references the previous dataset, e.g dataset 620 references 619:

```
  public CompiledDataset620(java.util.Map arguments) {
    field0 = ((org.logstash.generated.CompiledDataset619) arguments.get("field0"));
    field1 = ((org.jruby.RubyArray) arguments.get("field1"));
    field2 = ((org.logstash.config.ir.compiler.OutputDelegatorExt) arguments.get("field2"));
  }
```

So even if `CompiledDataset602` is scrubbed from the code text, the 619 is still there.

Reply:
agree, timing/performance related tests should be better done in integration tests.

Reply:
we can't scrub that because if we so then we loose callsite for the previous filter

Reply:
!!! @jsvd  -- I think your comment above about referencing previous datasets has knocked something loose in my brain. 

I'm pretty confident that we _don't_ need to reference previous datasets, which would make units significantly more cacheable and potentially eliminate the need for the global lock. Preliminary tests with the below generates code that _doesn't_ reference specific datasets, and seems to work, at least with moderate complexity if/else chaining.

~~~ diff
diff --git a/logstash-core/src/main/java/org/logstash/config/ir/compiler/VariableDefinition.java b/logstash-core/src/main/java/org/logstash/config/ir/compiler/VariableDefinition.java
index e6644cca6..83013003b 100644
--- a/logstash-core/src/main/java/org/logstash/config/ir/compiler/VariableDefinition.java
+++ b/logstash-core/src/main/java/org/logstash/config/ir/compiler/VariableDefinition.java
@@ -72,6 +72,8 @@ final class VariableDefinition implements SyntaxElement {
             safe = EventCondition.class;
         } else if (DynamicMethod.class.isAssignableFrom(clazz)) {
             safe = DynamicMethod.class;
+        } else if (BaseDataset.class.isAssignableFrom(clazz)) {
+            safe = BaseDataset.class;
         } else {
             safe = clazz;
         }
~~~